exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -0,0 +1,241 @@
+   * built-in service such as the time oracle.
+   *
+   * @return the service instance or null if there is no service with such id
+   * @throws IllegalArgumentException if the service with given id was not found or could not be
+   *     cast to given class
+   */
+  public <T extends Service> T getService(short serviceId, Class<T> serviceClass) {
+    Service service = services.get(serviceId);
+    checkArgument(service != null, "Service with given id=%s was not found", serviceId);
+    checkArgument(service.getClass().equals(serviceClass),
+        "Service (id=%s, class=%s) cannot be cast to %s",
+        serviceId, service.getClass().getCanonicalName(), serviceClass.getCanonicalName());
+    return serviceClass.cast(service);
+  }
+
+  private native long nativeCreateTestKit(UserServiceAdapter[] services, boolean auditor,
+      short withValidatorCount, TimeProvider timeProvider);
+
+  private native long nativeCreateSnapshot(long nativeHandle);
+
+  private native Block nativeCreateBlock(long nativeHandle);
+
+  private native Block nativeCreateBlockWithTransactions(long nativeHandle, byte[][] transactions);
+
+  private native EmulatedNode nativeGetEmulatedNode(long nativeHandle);
+
+  /**
+   * Creates a new builder for the TestKit.
+   *
+   * @param nodeType type of the main TestKit node - either validator or auditor. Note that
+   * {@link Service#afterCommit(BlockCommittedEvent)} logic will only be called on the main TestKit
+   *     node of this type
+   */
+  public static Builder builder(EmulatedNodeType nodeType) {
+    checkNotNull(nodeType);
+    return new Builder(nodeType);
+  }
+
+  /**
+   * Builder for the TestKit.
+   */
+  public static final class Builder {
+
+    private EmulatedNodeType nodeType;
+    private short validatorCount;
+    private List<Class<? extends ServiceModule>> services = new ArrayList<>();
+    private TimeProvider timeProvider;
+
+    private Builder(EmulatedNodeType nodeType) {
+      // TestKit network should have at least one validator node
+      if (nodeType == EmulatedNodeType.AUDITOR) {
+        validatorCount = 1;
+      }
+      this.nodeType = nodeType;
+    }
+
+    /**
+     * Sets number of additional validator nodes in the TestKit network. Note that
+     * regardless of the configured number of validators, only a single service will be
+     * instantiated.
+     */
+    public final Builder withValidators(short validatorCount) {
+      this.validatorCount = validatorCount;
+      return this;
+    }
+
+    /**
+     * Adds a service with which the TestKit would be instantiated. Several services can be added.
+     */
+    public final Builder withService(Class<? extends ServiceModule> serviceModule) {
+      services.add(serviceModule);
+      return this;
+    }
+
+    /**
+     * Adds services with which the TestKit would be instantiated.
+     */
+    @SafeVarargs
+    public final Builder withServices(Class<? extends ServiceModule> serviceModule,
```

Comment:
actually, `serviceModule` param can be removed and `java.util.Arrays#asList` is used as the varargs wrapper. Or did I misunderstand something?)

Reply:
> a requirement of at least one service module
There is `withService` in this case.

Reply:
Are there any disadvantages to the present signature? If not, I'd leave it as it is